File: 01_DFS/JUNG/boj1012.cpp
#include <bits/stdc++.h>
using namespace std;

int dx[4] = {-1, 1, 0, 0};
int dy[4] = {0, 0, -1, 1};

bool bfs(auto& graph, int m, int n, int x, int y) {
    if (graph[x][y] == 0)
        return false;

    queue<pair<int, int>> que;
    graph[x][y] = 0;
    que.push(pair<int, int>(x, y));

    while (!que.empty()) {
        auto front = que.front();
        que.pop();
        for (int i = 0; i < 4; i++) {
            int nx = front.first + dx[i];
            int ny = front.second + dy[i];
            if (nx <= -1 || ny <= -1 || nx >= m || ny >= n) continue;
            if (graph[nx][ny] == 1) {
                graph[nx][ny] = 0;
                que.push(pair<int, int>(nx, ny));
            }
        }
    }
    return true;
}

int main(void) {
    int T;
    cin >> T;

    while (T--) {
        int m, n, k, x, y, count = 0;
        cin >> m >> n >> k;
        vector<vector<int>> graph(m, vector<int>(n));

        for (int i = 0; i < k; i++) {
            cin >> x >> y;
            graph[x][y] = 1;
        }

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (bfs(graph, m, n, i, j) == true) count++;
            }
        }
        cout << count << "\n";
    }
    return 0;
}

File: 01_DFS/JUNG/boj1012.py
from collections import deque

# 테스트 케이스의 개수 t
t = int(input())
# 가로길이 m, 세로길이 n, 배추가 심어져 있는 위치의 개수 k
m, n, k = 0, 0, 0
graph = []

dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]


def bfs(graph, x, y, m, n):
    if graph[x][y] == 0:
        return False

    queue = deque()

    queue.append((x, y))
    graph[x][y] = 0
    while queue:
        front = queue.popleft()
        for i in range(4):
            nx = front[0] + dx[i]
            ny = front[1] + dy[i]
            if(nx <= -1 or ny <= -1 or nx >= m or ny >= n):
                continue
            if graph[nx][ny] == 1:
                queue.append((nx, ny))
                graph[nx][ny] = 0
    return True


while (t > 0):
    t -= 1

    m, n, k = map(int, input().split())
    for _ in range(m):
        graph.append([0 for _ in range(n)])

    for _ in range(k):
        i, j = map(int, input().split())
        graph[i][j] = 1

    count = 0
    for i in range(m):
        for j in range(n):
            if bfs(graph, i, j, m, n) == True:
                count += 1
    print(count)
    graph = []
